File: algorithms/0242.Valid Anagram/solution.rb
# @param {String} s
# @param {String} t
# @return {Boolean}
def is_anagram(s, t)
  if s.length != t.length then
    return false
  end
  map = {}
  s.each_char do |c|
    map[c] = map[c] || 0
    map[c] += 1
  end
  t.each_char do |c|
    if !map[c] || map[c] == 0 then
      return false
    end
    map[c] -= 1
  end
  return true
end

File: algorithms/0242.Valid Anagram/solution.rs
pub struct Solution;

const CHAR_SIZE: usize = 26;

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        Solution::count(s) == Solution::count(t)
    }

    fn count(s: String) -> [i32; CHAR_SIZE] {
        let mut counts = [0; CHAR_SIZE];
        for c in s.bytes() {
            counts[(c - b'a') as usize] += 1;
        }
        counts
    }
}

#[test]
fn test1() {
    assert_eq!(
        Solution::is_anagram(String::from("anagram"), String::from("nagaram"),),
        true
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::is_anagram(String::from("rat"), String::from("car"),),
        false
    );
}
